File: java-algo/src/main/java/com/leetcode/Problem382.java
package com.leetcode;

import com.common.ListNode;

import java.util.Random;

/**
 * 382. Linked List Random Node
 * https://leetcode.com/problems/linked-list-random-node/
 */
public class Problem382 {

    private final ListNode head;
    private final Random random;

    public Problem382(ListNode head) {
        this.head = head;
        this.random = new Random();
    }

    public int getRandom() {
        int scope = 1;
        int chosen = 0;
        ListNode cur = this.head;
        while (cur != null) {
            if (this.random.nextInt(scope) == 0) {
                chosen = cur.val;
            }
            cur = cur.next;
            scope++;
        }
        return chosen;
    }
}

File: python-algo/leetcode/problem_382.py
"""382. Linked List Random Node
https://leetcode.com/problems/linked-list-random-node/
"""
import random
from typing import Optional

from common.list_node import ListNode


class Solution:

    def __init__(self, head: Optional[ListNode]):
        self.head = head

    def get_random(self) -> int:
        scope = 1
        chosen_val = 0
        cur = self.head
        while cur:
            if random.random() < 1 / scope:
                chosen_val = cur.val
            cur = cur.next
            scope += 1
        return chosen_val
